### DISCUSSION TITLE
How to Debug JSBSim with UE5?

### DISCUSSION THREAD

**Discussion by gallonmate** [JSBSim-Team/jsbsim#761] [Q&A]
So I'm trying to get Debugging working for JSBSim as part of the UE5 plugin. I can get the Symbols loaded for the JSBSim.dll but it crashes during the first stages of initializing JSBSim.  Release build works fine with UE5, it is just debug version that causes some strange crashes. I will summarize the experience and info below:

@AlbanBERGERET-Epic 
Suggests that Debug builds should work, as mentioned here. https://github.com/JSBSim-Team/jsbsim/discussions/623#discussioncomment-2759637

But first the debug symbols and version of JSBSim don't actually seem to get built with UE5 DebugGame_Editor. So I make a quick edit to the UE5 project's JSBSim.Build.cs file, and just comment out the line below to force debug build.
```
		//if (Target.Configuration == UnrealTargetConfiguration.Debug
```
After this JSBSim.dll is built and loads debug symbols. UE5 starts and JSBSim can begin to initialize. 
![image](https://user-images.githubusercontent.com/28702753/200251457-7cb9e5e4-70af-4359-b684-26cb922c92ac.png)
In the above picture UE5 calls the various set paths. Everything is fine until it reaches SetSystemsPath.

This is then called:
![image](https://user-images.githubusercontent.com/28702753/200252254-ff367075-20f4-4530-a5eb-c8b262b4195d.png)

And then this is called and it crashes:
![image](https://user-images.githubusercontent.com/28702753/200252509-407c8b09-f3ee-4c9b-bcb4-dba3b89fb251.png)

I'm not sure I really understand why there is some error with strings when in release build everything works fine. Is there some debug build settings that changes some compiler settings on how string conversion are handled? 

Interestingly, if build the UE5 as Development instead of Debug but keep JSBSim as debug, I can get debug asserts that happen starting before the crash, starting with SetAircraftPath. I can ignore these asserts until it finally crashes as the same point as shown in the snippets above.
![image](https://user-images.githubusercontent.com/28702753/200253812-87a84697-3835-425d-a9e3-5c24c14bf355.png)
![image](https://user-images.githubusercontent.com/28702753/200254547-0dc3ab1e-e178-4925-82dc-6494c68660f6.png)

**Comment by seanmcleod** [comment]
I wonder if there is some mismatch between the JSBSim module and the UE module in terms of the assumption regarding Character Set (Unicode or not). In particular possibly a different setting for Character Set between the debug and release builds of JSBSim?

![image](https://user-images.githubusercontent.com/824856/200259756-afc7286a-cde2-478e-ae48-6c2eda08b610.png)

**Reply by gallonmate** [reply]
Thanks Sean! I tried changing the character set for the JSBSim build and no dice. Unfortunately there's no property I can change for the Unreal engine solution, 90% of properties are not available, due to how I guess Unreal build works.

I did end up finding a work around. Not Ideal but it works. After trying a bunch of different configurations with the Debug build, nothing helped. So I ended up just changing the JSBSim Release build to generate debug info, in the project properties Linker, Debugging. Of course still have to change some other settings so code/variables aren't optimized away but at least I can debug and breakpoint now in JSBSim through Unreal.

![image](https://user-images.githubusercontent.com/28702753/200448111-fc84745a-df7f-4e0f-9c67-6d2e53ddf83b.png)

**Comment by seanmcleod** [comment]
What do you see if you start stepping through the code from the call to `Exec->SetSystemsPath()`? Watch what happens to the string contents as the `FString` is converted to a `std::wstring` etc. and as that is passed on. Keep an eye on whether it's Unicode or not.

Looks like a corrupted string in a couple of your screenshots.

![image](https://user-images.githubusercontent.com/824856/200509783-26dcff74-13ba-4802-be67-f1846bdd3de5.png)

![image](https://user-images.githubusercontent.com/824856/200509909-82476fd3-b7ce-42ea-937e-c5f1f071138a.png)

**Reply by AlbanBERGERET-Epic** [reply]
And be careful to properly convert strings when you use specific encodings. There are some useful macros explained there: 
https://docs.unrealengine.com/5.0/en-US/character-encoding-in-unreal-engine/

**Comment by gallonmate** [comment]
Sean, I've stepped though it all a few times but it's mostly gibberish to me, as I'm an amateur with programming. But I'm not really worried anymore, as the JSBSim debugging works in a release build. And that's good enough :smile: 

Alban, no strings have been altered. The plugin is unchanged from the original. It crashes simple by using the JBSim debug dll. Release dll works fine. Although I am using a Binary build of Unreal and not a Source build of Unreal. After digging around, I'm starting to think it's simply the debug dll that might not be compatible with a regular Unreal binary build (downloaded though the epic launcher).

Interestingly I can get the JSBSim Release build to produce the exact same crash if set the JSBsim runtime library to debug dll, with no other changes made. So I'm thinking a debug dll is just incompatible with the a stock binary build of the Unreal Editor. And would probably work with a source build of Unreal.
![image](https://user-images.githubusercontent.com/28702753/200748899-275c3f60-4242-4a68-84ed-13883a30fc46.png)

**Comment by seanmcleod** [comment]
Hmm, certain combinations of the C/C++ runtime between different modules (JSBSim and Unreal) can cause issues depending on what sorts of data is passed between the modules.

So there is a good chance that Unreal has been built with the release DLL version of the runtime, i.e. **Multi-threaded DLL (/MD)** and that's causing issues when some data is passed into JSBSim which is built with the debug DLL version of the runtime when you generate a debug build of JSBSim, i.e. **Multi-threaded Debug DLL (/MDd)**.

One option to try is to switch JSBSim to using the static rather than DLL versions of the runtime, i.e. **Multi-threaded Debug (/MTd)** for the debug build of JSBSim and **Multi-threaded (/MT)** for the release build of JSBSim.

**Reply by gallonmate** [reply]
Unfortunately same bug trying that. Although I also realized using jsbsim as a dll is sort of new to this repo, previously it was just exe or static lib. It was only added this summer as an option to the JSBSim repo. Regardless, the jsbsim debugging as a release build works plenty well with UE5, so I'm dropping further attempts on my end. Thanks again for the time & thoughts!
